crepecake/src/main/java/com/google/cloud/tools/crepecake/registry/ManifestPuller.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.crepecake.registry;
+
+import com.google.api.client.http.HttpResponseException;
+import com.google.cloud.tools.crepecake.blob.Blobs;
+import com.google.cloud.tools.crepecake.http.Authorization;
+import com.google.cloud.tools.crepecake.http.Connection;
+import com.google.cloud.tools.crepecake.http.Request;
+import com.google.cloud.tools.crepecake.http.Response;
+import com.google.cloud.tools.crepecake.image.json.ManifestTemplate;
+import com.google.cloud.tools.crepecake.image.json.UnknownManifestFormatException;
+import com.google.cloud.tools.crepecake.image.json.V21ManifestTemplate;
+import com.google.cloud.tools.crepecake.image.json.V22ManifestTemplate;
+import com.google.cloud.tools.crepecake.json.JsonTemplateMapper;
+import com.google.cloud.tools.crepecake.registry.json.ErrorEntryTemplate;
+import com.google.cloud.tools.crepecake.registry.json.ErrorResponseTemplate;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+import javax.annotation.Nullable;
+
+/** Registry interface for pulling an image's manifest. */
+public class ManifestPuller {
+
+  // TODO: This should be configurable.
+  private static final String PROTOCOL = "http";
+
+  @Nullable private final Authorization authorization;
+  private final String serverUrl;
+  private final String baseImage;
+
+  /**
+   * Instantiates a {@link ManifestTemplate} from a JSON string. This checks the {@code
+   * schemaVersion} field of the JSON to determine which manifest version to use.
+   */
+  public static ManifestTemplate getManifestTemplateFromJson(String jsonString)
+      throws IOException, UnknownManifestFormatException {
+    ManifestTemplate manifestTemplate =
+        JsonTemplateMapper.readJson(jsonString, ManifestTemplate.class);
+
+    switch (manifestTemplate.getSchemaVersion()) {
```

Comment:
could this be pushed into the ManifestTemplate class to avoid the switch statement?

Reply:
`V21ManifestTemplate` and `V22ManifestTemplate` are subclasses of `ManifestTemplate`. Perhaps I should make another class called `ManifestTemplates` with a `#fromJson` method that does this?

Reply:
What you're doing here is close to 


switch (type) {
  case type1:
   doThing1
  case Type2:
    doThing2

This pattern is usually a failure to apply OO. The OO equivalent is

type.doThing()

In this case what seems called for is

manifestTemplate.readJson()

If the `manifestTemplate` variable is a V21ManifestTemplate then it will use that. If it's V22ManifestTemplate then it uses that.

Reply:
I'm probably going about this the wrong way but it seems that in this situation, polymorphism doesn't apply since it is first parsing the JSON as a `ManifestTemplate` object and then re-parsing the JSON as one of the subclasses according to a field in the `ManifestTemplate` (not resolved to one of the subclasses yet).

Reply:
That is quite weird and unconventional. It may suggest that data binding is not the best way to read these files.

Reply:
Should I do something like described in https://stackoverflow.com/questions/26190851/get-single-field-from-json-using-jackson?

Reply:
That's one way of doing it, I guess.